### PR TITLE
Validate the OPCODE_SIGNATURES table

### PR DESCRIPTION
**Pull request by elliottt**
Add a test to cranelift-fuzzgen that validates the `OPCODE_SIGNATURES` table by comparing it against valid instruction instantiations. Additionally, fix some signatures that were identified as not matching the signature from the instruction's specification.

This is incremental work towards generating instruction instantiations in cranelift-fuzzgen using the instruction specification instead of explicitly opting-in an instruction.

### CODE REVIEW

File: cranelift/fuzzgen/src/bin/check_opcode_signatures.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            args: args.as_ref(),
+            rets: rets.as_ref(),
+        });
+    }
+
+    insts
+}
+
+fn main() {
+    let all_ops = instruction_instantiations();
```
Comment (elliottt) [inline cranelift/fuzzgen/src/bin/check_opcode_signatures.rs:114]: An alternate approach here would be to instead use the signature to check the `OPCODE_SIGNATURES` table, however explicitly generating all instantiations leaves the door open to using that set to generate the complement of signatures that we don't currently generate instructions for when fuzzing.

File: cranelift/fuzzgen/src/bin/check_opcode_signatures.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    let mut unknown = vec![];
+    for (op, insts) in fuzzed.iter() {
+        if let Some(known_insts) = everything.get(op) {
+            let invalid = Vec::from_iter(insts.iter().filter(|inst| !known_insts.contains(inst)));
+            if !invalid.is_empty() {
+                println!("# Invalid instantiations for Opcode::{:?}", op);
+                for inst in invalid {
+                    println!("- args: `{:?}`, rets: `{:?}`", inst.args, inst.rets);
+                }
+            }
```
Comment (elliottt) [inline cranelift/fuzzgen/src/bin/check_opcode_signatures.rs:127]: One question I have here is if this is the best way to report this information. Perhaps reworking this program as a test instead would be better, as that would prevent any backsliding on the existing signatures?

Comment (afonso360) [reply cranelift/fuzzgen/src/bin/check_opcode_signatures.rs:125]: Yeah, changing this to be a test would be awesome, that way we definitely wouldn't miss wrong signatures. (which there are way more than I would have thought 😅)

File: cranelift/fuzzgen/src/function_generator.rs [hunk 3]
```diff
@@ -1266,7 +1259,7 @@ const OPCODE_SIGNATURES: &[OpcodeSignature] = &[
     (Opcode::Fma, &[F32, F32, F32], &[F32]),
     (Opcode::Fma, &[F64, F64, F64], &[F64]),
     (Opcode::Fma, &[F32X4, F32X4, F32X4], &[F32X4]),
-    (Opcode::Fma, &[F64X2, F64X2, F64X4], &[F64X2]),
+    (Opcode::Fma, &[F64X2, F64X2, F64X2], &[F64X2]),
```
Comment (afonso360) [inline cranelift/fuzzgen/src/function_generator.rs:1262]: I was initially surprised as to how this didn't crash, but I guess we never generate `F64X4`'s so we never selected this!

File: cranelift/fuzzgen/src/bin/check_opcode_signatures.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                for current in argss.iter() {
+                    for ty in col {
+                        let mut args = vec![*ty];
+                        args.extend_from_slice(&current);
+                        next.push(args);
+                    }
+                }
+
+                let _ = std::mem::replace(&mut argss, next);
+            }
```
Comment (afonso360) [inline cranelift/fuzzgen/src/bin/check_opcode_signatures.rs:78]: This took me a while to figure out, but I think its the same thing I wrote in my [proof of concept code](https://github.com/afonso360/cranelift-opcode-gen/blob/75bb630898fef4761d8058dd5a5582364e301ca4/src/main.rs#L15-L37) a while ago!

Adding some sort of comment explaining this operation at a high level would be really helpful. 

In my POC I borrowed the name from [Itertools::multi_cartesian_product](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.multi_cartesian_product) which does the same operation, I'm not sure if that is the correct mathematical name for the operation, but it made sense at the time.

**Review by afonso360 — APPROVED** [review]
This is awesome! Just a small request for some comments since it took me quite a while to understand that part of the code.

I'm surprised at just how many instructions were wrong 😅 . But I'm really happy that whole thing is going away soon!

File: cranelift/fuzzgen/src/bin/check_opcode_signatures.rs (hunk 4 above)
Comment (elliottt) [reply cranelift/fuzzgen/src/bin/check_opcode_signatures.rs:78]: That's definitely the right name -- I'll add some comments so this thing isn't quite so opaque :)

File: cranelift/fuzzgen/src/bin/check_opcode_signatures.rs (hunk 2 above)
Comment (elliottt) [reply cranelift/fuzzgen/src/bin/check_opcode_signatures.rs]: I reworked it as a test, so you'll see the output if a new signature is added that doesn't match the constraints of the instruction DSL :+1:

File: cranelift/fuzzgen/src/function_generator.rs [hunk 5]
```diff
@@ -2347,3 +2340,191 @@ where
...
+        Opcode::AtomicRmw,
+        Opcode::Sload8,
+        Opcode::Sload16,
+        Opcode::Sload32,
+        Opcode::Uload8,
+        Opcode::Uload16,
+        Opcode::Uload32,
+        Opcode::Istore8,
+        Opcode::Istore16,
+        Opcode::Istore32,
```
Comment (elliottt) [inline cranelift/fuzzgen/src/function_generator.rs:2384]: This is the current list of discrepancies that we'll need to fix to ensure that `OPCODE_SIGNATURES` is reflecting the current state of instruction constraints. A separate issue will be to reconcile the instruction constraints with the current behavior of the backends, but we'll get there eventually :)

**Pull request merged**